wgpu/gui/events.py

```diff
@@ -0,0 +1,236 @@
+        is a dict with at least the key event_type. For details, see
+        https://jupyter-rfb.readthedocs.io/en/latest/events.html
+        """
+        event_type = event.type
+        for callback in self._event_handlers[event_type]:
+            callback(event)
+
+    def add_event_handler(self, *args):
+        """Register an event handler.
+
+        Arguments:
+            callback (callable): The event handler. Must accept a
+                single event argument.
+            *types (list of strings): A list of event types.
+
+        For the available events, see
+        https://jupyter-rfb.readthedocs.io/en/latest/events.html
+
+        Can also be used as a decorator.
+
+        Example:
+
+        .. code-block:: py
+
+            def my_handler(event):
+                print(event)
+
+            canvas.add_event_handler(my_handler, "pointer_up", "pointer_down")
+
+        Decorator usage example:
+
+        .. code-block:: py
+
+            @canvas.add_event_handler("pointer_up", "pointer_down")
+            def my_handler(event):
+                print(event)
+        """
+        decorating = not callable(args[0])
+        callback = None if decorating else args[0]
+        types = args if decorating else args[1:]
+
+        def decorator(_callback):
+            for type in types:
+                self._event_handlers[type].add(_callback)
+            return _callback
+
+        if decorating:
+            return decorator
+        return decorator(callback)
+
+    def remove_event_handler(self, callback, *types):
+        """Unregister an event handler.
+
+        Arguments:
+            callback (callable): The event handler.
+            *types (list of strings): A list of event types.
+        """
+        for type in types:
+            self._event_handlers[type].remove(callback)
+
+
+class EventDispatcher(EventTarget):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def handle_event(self, event: Event):
+        target = event.target
+        while target and target is not self:
+            pointer_id = getattr(event, "pointer_id", None)
+            if pointer_id and pointer_id in EventTarget.pointer_captures:
+                capture = EventTarget.pointer_captures[pointer_id]
+                capture.handle_event(event)
+                # Encountered an event with pointer_id while there is a
+                # capture active, so don't bubble, just return immediately
+                return
+            else:
+                target.handle_event(event)
+                if pointer_id and pointer_id in EventTarget.pointer_captures:
+                    # Apparently the set_pointer_capture was called for this
+                    # event.pointer_id, so return immediately
+                    return
+            if not event.bubbles:
+                break
+            target = getattr(target, "parent", None)
```

Comment:
This looks pygfx-agnostic at first glance, but it does assume the name 'parent' ... is this a sign that we're too opinionated here?